frame/vesting/src/lib.rs

```diff
@@ -301,116 +362,383 @@ pub mod pallet {
+
+		let per_block = {
+			let duration = ending_block
+				.saturating_sub(T::BlockNumberToBalance::convert(starting_block))
+				.max(One::one());
+			(locked / duration).max(One::one())
+		};
+
+		// While `per_block` should never be 0, it is possible that the created schedule would end
+		// after the highest possible block, which is a case we are ok with, but `validate` fails.
+		let schedule = VestingInfo::new::<T>(locked, per_block, starting_block);
+
+		Ok(Some(schedule))
+	}
+
+	// Execute a vested transfer from `source` to `target` with the given `schedule`.
+	fn do_vested_transfer(
+		source: <T::Lookup as StaticLookup>::Source,
+		target: <T::Lookup as StaticLookup>::Source,
+		schedule: VestingInfo<BalanceOf<T>, T::BlockNumber>,
+	) -> DispatchResult {
+		// Validate user inputs.
+		ensure!(schedule.locked() >= T::MinVestedTransfer::get(), Error::<T>::AmountLow);
+		schedule.validate::<T::BlockNumberToBalance, T>()?;
+
+		let target = T::Lookup::lookup(target)?;
+		let source = T::Lookup::lookup(source)?;
+
+		// Check we can add to this account prior to any storage writes.
+		Self::can_add_vesting_schedule(
+			&target,
+			schedule.locked(),
+			schedule.per_block(),
+			schedule.starting_block(),
+		)?;
+
+		T::Currency::transfer(
+			&source,
+			&target,
+			schedule.locked(),
+			ExistenceRequirement::AllowDeath,
+		)?;
+
+		// We can't let this fail because the currency transfer has already happened.
+		let res = Self::add_vesting_schedule(
+			&target,
+			schedule.locked(),
+			schedule.per_block(),
+			schedule.starting_block(),
+		);
+		debug_assert!(res.is_ok(), "Failed to add a schedule when we had to succeed.");
+
+		Ok(())
+	}
+
+	/// Iterate through the schedules to track the current locked amount and
+	/// filter out completed and specified schedules.
+	///
+	/// Returns a tuple that consists of:
+	/// - vec of vesting schedules, where completed schedules and those
+	/// 	specified by filter are remove. (Note the vec is not checked for respecting
+	/// 	bounded length.)
+	/// - the amount locked at the current block number based on the given schedules.
+	///
+	/// NOTE: the amount locked does not include any schedules that are filtered out.
+	fn report_schedule_updates(
+		schedules: Vec<VestingInfo<BalanceOf<T>, T::BlockNumber>>,
+		action: VestingAction,
+	) -> (Vec<VestingInfo<BalanceOf<T>, T::BlockNumber>>, BalanceOf<T>) {
 		let now = <frame_system::Pallet<T>>::block_number();
-		let locked_now = vesting.locked_at::<T::BlockNumberToBalance>(now);
 
-		if locked_now.is_zero() {
-			T::Currency::remove_lock(VESTING_ID, &who);
-			Vesting::<T>::remove(&who);
-			Self::deposit_event(Event::<T>::VestingCompleted(who));
+		let mut total_locked_now: BalanceOf<T> = Zero::zero();
+		let filtered_schedules = schedules
+			.into_iter()
+			.enumerate()
+			.filter_map(|(index, schedule)| {
+				let locked_now = schedule.locked_at::<T::BlockNumberToBalance>(now);
+				if locked_now.is_zero() || action.should_remove(index) {
```

Comment:
```suggestion
				if locked_now.is_zero() {
```

I am still of the opinion that this function is doing 2 things :D

Reply:
I agree, but this way we can do just 1 iteration to get the updated lock and remove old schedules instead of 2 separate iterations.

Additionally, I can't think of a scenario where you would ever want to update lock but not remove finishing schedules and vice versa.

Still open to suggestions though since there probably is a better way to organize this code, but I can't think of it :D

Reply:
I have: 45726f764710da36829fb3ef131d48799b4d30d0 

We could get fancier by
- `pick_schedules` taking an Iter as arg instead of a vec, but I was having trouble with this and seemed like over engineering
- making a `pick_locked` function that takes an Iter and returns only schedules with some locked funds, but this would mean we would have to call `locked_at` twice when we have another map to increment the total locked sum

If there where a way to create a function took/return an Iter and also somehow incremented a locked sum from its callsite without taking ownership, we could turn the `filter_map` into a reusable function

Reply:
Tried something like below, but the lifetimes get tricky because `total_locked_now` is only incremented why the Iter is consumed

```
	fn track_locked(schedules: impl Iterator<Item = VestingInfo<BalanceOf<T>, T::BlockNumber>> ) -> (impl Iterator<Item = VestingInfo<BalanceOf<T>, T::BlockNumber>>, BalanceOf<T>) {
		let now = <frame_system::Pallet<T>>::block_number();
		let mut total_locked_now: BalanceOf<T> = Zero::zero();
		let locked_schedules = schedules.filter_map(|schedule| {
			let locked_now = schedule.locked_at::<T::BlockNumberToBalance>(now);
			if locked_now.is_zero() {
				None
			} else {
				total_locked_now = total_locked_now.saturating_add(locked_now);
				Some(schedule)
			}
		});

		(locked_schedules, total_locked_now)
	}

Reply:
You'd probably have an easier time with lifetimes if you avoid `impl Trait` and use `Box<dyn Trait>`.

Either way, I agree that my comment was somewhat opinionated. I'm glad you re-designed a few things, but no need to furhter revamp.

Reply:
Hmm, I was avoiding`dyn Trait` because I thought it was a performance hit because it is chosen at runtime vs `impl Trait` which is made concrete at compile time